mlfinlab/feature_importance/orthogonal.py

```diff
@@ -0,0 +1,110 @@
+
+    :param data_frame: (pd.DataFrame): to standardize
+    :return: (pd.DataFrame): standardized data frame
+    """
+    return data_frame.sub(data_frame.mean(), axis=1).div(data_frame.std(), axis=1)
+
+
+def get_orthogonal_features(feature_df, variance_thresh=.95):
+    """
+    Snippet 8.5, page 119. Computation of Orthogonal Features.
+
+    Get PCA orthogonal features
+
+    :param feature_df: (pd.DataFrame): with features
+    :param variance_thresh: (float): % of overall variance which compressed vectors should explain
+    :return: (pd.DataFrame): compressed PCA features which explain %variance_thresh of variance
+    """
+    # Given a dataframe of features, compute orthofeatures
+    feature_df_standard = _standardize_df(feature_df)  # Standardize
+    dot_matrix = pd.DataFrame(np.dot(feature_df_standard.T, feature_df_standard), index=feature_df.columns,
+                              columns=feature_df.columns)
+    _, eigen_vec = _get_eigen_vector(dot_matrix, variance_thresh)
+    pca_features = np.dot(feature_df_standard, eigen_vec)
+    return pca_features
+
+
+def get_pca_rank_weighted_kendall_tau(feature_imp, pca_rank):
+    """
+    Snippet 8.6, page 121. Computation of Weighted Kendall's Tau Between Feature Importance and Inverse PCA Ranking
+
+    :param feature_imp: (np.array): with feature mean importance
+    :param pca_rank: (np.array): PCA based feature importance rank
+    :return: (float): weighted Kendall tau of feature importance and inverse PCA rank with p_value
+    """
+    return weightedtau(feature_imp, pca_rank ** -1.)
+
+
+def feature_pca_analysis(feature_df, feature_importance, variance_thresh=0.95):
+    """
+    Perform correlation analysis between feature importance (MDI for example, supervised)
+    and PCA eigen values (unsupervised). High correlation means that probably the pattern identified
+    by the ML algorithm is not entirely overfit.
+
+    :param feature_df: (pd.DataFrame): with features
+    :param feature_importance: (pd.DataFrame): individual MDI feature importance
+    :param variance_thresh: (float): % of overall variance which compressed vectors should explain in PCA compression
+    :return: (dict): with kendall, spearman, pearson and weighted_kendall correlations and p_values
+    """
+    feature_df_standard = _standardize_df(feature_df)  # Standardize
+    dot = pd.DataFrame(np.dot(feature_df_standard.T, feature_df_standard), index=feature_df.columns,
+                       columns=feature_df.columns)
+    eigen_val, eigen_vec = _get_eigen_vector(dot, variance_thresh)
+
+    # Compute correlations between eigen values for each eigen vector vs mdi importance
+    all_eigen_values = []  # All eigen values in eigen vectors
+    corr_dict = {'Pearson': [], 'Spearman': [], 'Kendall': []}  # Dictionary containing correlation metrics
+    for vec in eigen_vec.columns:
+        all_eigen_values.extend(eigen_vec[vec].values * eigen_val[vec])
```

Comment:
The computation of the orthogonal features can be computed in a less error prone way using the Skearn implementation of the PCA algorithm. I don't see a reason for this re-implementation. The SKlearn interface to the algorithm also has the nice `Transformer` interface. You can check it out here https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html

Reply:
sklearn's method centered input data, but doesn't scale, that is why I decided to keep it as it is from the book

Reply:
Alright, let's leave it as it is in the book then. The code looks correct to me otherwise.